server/server.js

```diff
@@ -12,18 +29,487 @@ const db = new pg.Pool({
+        const results = await db.query(sql, params);
+        let data = results.rows;
+        data.forEach((entry) => {
+          entry.year = dayjs.tz(entry.date).year();
+          entry.week = dayjs.tz(entry.date).week();
+          entry.activeYear = dayjs.tz(entry.activeDate).year();
+          entry.activeWeek = dayjs.tz(entry.activeDate).week();
+        });
+        data = data.filter((entry) => (
+          (entry.activeYear === currentYear ? entry.activeWeek !== currentWeek : true)
+        ));
+        const tracker = {
+          groups: [],
+          penalties: []
+        };
+        for (let d = 0; d < data.length; d++) {
+          const { keyGroupId, groupName, frequencyReq, intervalReq, year, week } = data[d];
+          if (tracker.groups.indexOf(groupName) === -1) {
+            tracker.groups.push(groupName);
+            tracker[groupName] = { id: keyGroupId, frequency: frequencyReq, interval: intervalReq, count: 0 };
+          }
+          if (year === currentYear && week === lastWeek) {
+            tracker[groupName].count++;
+          }
+        }
+        for (let g = 0; g < tracker.groups.length; g++) {
+          const currGroup = tracker[tracker.groups[g]];
+          if (currGroup.count < currGroup.frequency) {
+            tracker.penalties.push(currGroup.id);
+          }
+        }
+        const sql2 = `
+          SELECT "penaltyId"
+          FROM "penalties"
+        `;
+        const result2 = await db.query(sql2);
+        const penaltyIds = result2.rows.map(id => id.penaltyId);
+        if (penaltyIds.length) {
+          const totalPenalties = tracker.penalties.length;
+          if (totalPenalties) {
+            for (let p = 0; p < totalPenalties; p++) {
+              tracker.penalties = tracker.penalties.filter((penalty) => penaltyIds.indexOf(String(penalty).concat(user, currentWeek, currentYear)) === -1);
+            }
+          }
+        }
+        let sql3 = `
+          INSERT INTO "penalties" ("groupId", "userId", "week", "year")
+          VALUES
+          `;
+        if (tracker.penalties.length) {
+          for (let p = 0; p < tracker.penalties.length; p++) {
+            if (p !== tracker.penalties.length - 1) {
+              sql3 = sql3.concat(`($${p + 4}, $1, $2, $3), `);
+            } else {
+              sql3 = sql3.concat(`($${p + 4}, $1, $2, $3) `);
+            }
+          }
+          sql3 = sql3.concat('RETURNING *');
+          const params3 = [user, currentWeek, currentYear, ...tracker.penalties];
+          await db.query(sql3, params3);
+        }
+        const sql4 = `
+          SELECT "groups"."groupName" AS "groupName",
+            "date",
+            "status",
+            "groups"."betAmount" AS "betAmount",
+            "penaltyId"
+          FROM "penalties"
+          JOIN "groups" USING ("groupId")
+          WHERE "userId" = $1;
+        `;
+        const params4 = [user];
+        await db.query(sql4, params4);
+      });
+    }
```

Comment:
There's a lot of impressive logic in here. But it's too long for one function. Try extracting smaller functions (or the SQL) to make this more readable.